contracts/SP3QSAJQ4EA8WXEDSRRKMZZ29NH91VZ6C5X88FGZQ.crashpunks-v1.clar

```diff
@@ -0,0 +1,1004 @@
+        ;; u2 means bidding is in progress and the sale data can't be changed.
+        (asserts! (not (and (> currentAmount u0) (is-eq saleType u2))) bidding-error)
+        ;; owner or approval can do this.
+        (asserts! (unwrap! (is-approved nftIndex owner) nft-not-owned-err) nft-not-owned-err)
+        ;; Note - don't override the sale cyle index here as this is a public method and can be called ad hoc. Sale cycle is update at end of sale!
+        (asserts! (map-set nft-sale-data {nft-index: nftIndex} {sale-cycle-index: saleCycleIndex, sale-type: sale-type, increment-stx: increment-stx, reserve-stx: reserve-stx, amount-stx: amount-stx, bidding-end-time: bidding-end-time}) not-allowed)
+        (print {evt: "set-sale-data", nftIndex: nftIndex, saleType: sale-type, increment: increment-stx, reserve: reserve-stx, amount: amount-stx, biddingEndTime: bidding-end-time})
+        (ok true)
+    )
+)
+
+;; see nft-tradable-trait
+(define-public (unlist-item (nftIndex uint))
+    (let
+        (
+            (saleCycleIndex (unwrap! (get sale-cycle-index (map-get? nft-sale-data {nft-index: nftIndex})) amount-not-set))
+            (owner           (unwrap! (nft-get-owner? crashpunks nftIndex) not-allowed))
+        )
+        (asserts! (unwrap! (is-approved nftIndex owner) nft-not-owned-err) nft-not-owned-err)
+        ;; Note - don't override the sale cyle index here as this is a public method and can be called ad hoc. Sale cycle is update at end of sale!
+        (asserts! (map-set nft-sale-data {nft-index: nftIndex} {sale-cycle-index: saleCycleIndex, sale-type: u0, increment-stx: u0, reserve-stx: u0, amount-stx: u0, bidding-end-time: u0}) not-allowed)
+        (print {evt: "unlist-item", nftIndex: nftIndex})
+        (ok true)
+    )
+)
+
+;; see nft-tradable-trait
+(define-public (list-item (nftIndex uint) (amount uint))
+    (let
+        (
+            (saleCycleIndex (unwrap! (get sale-cycle-index (map-get? nft-sale-data {nft-index: nftIndex})) amount-not-set))
+            (owner           (unwrap! (nft-get-owner? crashpunks nftIndex) not-allowed))
+        )
+        (asserts! (unwrap! (is-approved nftIndex owner) nft-not-owned-err) nft-not-owned-err)
+        ;; (map-set approvals {owner: tx-sender, operator: operator, nft-index: token-id} true)
+        ;; Note - don't override the sale cyle index here as this is a public method and can be called ad hoc. Sale cycle is update at end of sale!
+        (asserts! (map-set nft-sale-data {nft-index: nftIndex} {sale-cycle-index: saleCycleIndex, sale-type: u1, increment-stx: u0, reserve-stx: u0, amount-stx: amount, bidding-end-time: u0}) not-allowed)
+        (print {evt: "list-item", nftIndex: nftIndex, amount: amount})
+        (ok true)
+    )
+)
+
+;; see nft-tradable-trait
+(define-public (buy-now (nftIndex uint) (owner principal) (recipient principal))
+    (let
+        (
+            (saleType (unwrap! (get sale-type (map-get? nft-sale-data {nft-index: nftIndex})) amount-not-set))
```

Comment:
extract `(unwrap! (map-get? nft-sale-data {nft-index: nftIndex}) err-not-found)` into a variable for better readability